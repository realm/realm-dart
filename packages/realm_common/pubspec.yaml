name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 3.2.0

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.3.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0

dev_dependencies:
  lints: ^3.0.0
