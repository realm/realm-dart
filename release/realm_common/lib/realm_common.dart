// Copyright 2021 MongoDB, Inc.
// SPDX-License-Identifier: Apache-2.0

export 'src/realm_common_base.dart';
export 'src/realm_types.dart';

export 'package:objectid/objectid.dart' show ObjectId;
export 'package:sane_uuid/uuid.dart' show Uuid;
