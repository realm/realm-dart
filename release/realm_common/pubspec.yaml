name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 2.0.0-alpha.5

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.0.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0-alpha.5

dev_dependencies:
  lints: ^3.0.0
