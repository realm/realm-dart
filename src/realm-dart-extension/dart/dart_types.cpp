////////////////////////////////////////////////////////////////////////////
//
// Copyright 2016 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////
#pragma clang diagnostic ignored "-Wdocumentation"

#include <cmath>
#include <functional>
#include <map>
#include <string>

#include "dart_api.h"
#include "dart_native_api.h"

#include "js_types.hpp"

#include "dart_types.hpp"

bool ___DartError::use_native_exceptions;

Dart_Handle operator ||(const Dart_Handle& dartResult, ___DartError error) {
	if (Dart_IsError(dartResult)) {
		auto err = Dart_GetError(dartResult);
		//Dart_PropagateError(dartResult);
		if (!___DartError::use_native_exceptions) {
			Dart_ThrowException(dartResult);
		}
		else {
			throw std::runtime_error(err);
		}
		
	}
	return dartResult;
}

namespace realm {
	namespace dartvm {
		namespace flutter {
			std::string FilesDir;
		}

		Dart_Handle DartCoreLibrary;
		Dart_Handle RealmLibrary;
		std::string RealmPackageName;
		Dart_Handle DartDateTimeType;
		
		Dart_Handle RealmCreateDateTimeFunc;
		Dart_Handle RealmDynamicObjectType;
		Dart_Handle RealmTypeStaticPropertiesType;
		Dart_Handle RealmHelpersType;
	}

	namespace util {
		void to_lower(char& c) {
			c = std::tolower(static_cast<unsigned char>(c));
		}

		void to_lowecase(std::string& string)
		{
			std::for_each(string.begin(), string.end(), to_lower);
		}
	}
}

namespace Dart {

	HandleScope::HandleScope() {
		Dart_EnterScope();
	}

	HandleScope::~HandleScope() {
		Dart_ExitScope();
	}

	Dart_Handle GetLibrary(const char* libName) {
		Dart_Handle libraryUrl = Dart_NewStringFromCString(libName) || handleError;
		Dart_Handle library = Dart_LookupLibrary(libraryUrl) || handleError;
		return library;
	}
}

