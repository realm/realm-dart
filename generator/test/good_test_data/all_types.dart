import 'dart:typed_data';

import 'package:realm_common/realm_common.dart';

@RealmModel()
@MapTo('MyFoo')
class _Foo {
  @Indexed()
  int x = 0;
  late _Bar? bar;
}

@RealmModel()
class _Bar {
  @PrimaryKey()
  late String name;
  @Indexed()
  late bool aBool, another; // both are indexed!
  var data = Uint8List(16);
  @MapTo('tidspunkt')
  @Indexed()
  var timestamp = DateTime.now();
  var aDouble = 0.0;
  // late Decimal128 decimal; // not supported yet
  _Foo? foo;
  @Indexed()
  late ObjectId objectId;
  @Indexed()
  late Uuid uuid;
  @Ignored()
  var theMeaningOfEverything = 42;
  var list = [0]; // list of ints with default value
  // late Set<int> set; // not supported yet
  // late map = <String, int>{}; // not supported yet

  @Indexed()
  String? anOptionalString;

  @Backlink(#bar)
  late Iterable<_Foo> foos;

  late RealmValue any;
  late List<RealmValue> manyAny;

  late Decimal128 decimal;
}

@RealmModel()
class _PrimitiveTypes {
  late String stringProp;
  late bool boolProp;
  late DateTime dateProp;
  late double doubleProp;
  late ObjectId objectIdProp;
}
