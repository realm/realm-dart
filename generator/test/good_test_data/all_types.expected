// **************************************************************************
// RealmObjectGenerator
// **************************************************************************

class Foo extends _Foo with RealmEntity, RealmObject {
  static var _defaultsSet = false;

  Foo({
    int x = 0,
  }) {
    if (!_defaultsSet) {
      _defaultsSet = RealmObject.setDefaults<Foo>({
        'x': 0,
      });
    }
    RealmObject.set(this, 'x', x);
  }

  Foo._();

  @override
  int get x => RealmObject.get<int>(this, 'x') as int;
  @override
  set x(int value) => RealmObject.set(this, 'x', value);

  @override
  Stream<RealmObjectChanges<Foo>> get changes =>
      RealmObject.getChanges<Foo>(this);

  @override
  Foo freeze() => RealmObject.freezeObject<Foo>(this);

  static SchemaObject get schema => _schema ??= _initSchema();
  static SchemaObject? _schema;
  static SchemaObject _initSchema() {
    RealmObject.registerFactory(Foo._);
    return const SchemaObject(Foo, 'MyFoo', [
      SchemaProperty('x', RealmPropertyType.int),
    ]);
  }
}

class Bar extends _Bar with RealmEntity, RealmObject {
  static var _defaultsSet = false;

  Bar(
    String id,
    bool aBool,
    bool another,
    ObjectId objectId, {
    Uint8List data = Uint8List(16),
    DateTime timestamp = DateTime.now(),
    double aDouble = 0.0,
    Foo? foo,
    String? anOptionalString,
    Iterable<int> list = const [],
  }) {
    if (!_defaultsSet) {
      _defaultsSet = RealmObject.setDefaults<Bar>({
        'data': Uint8List(16),
        'tidspunkt': DateTime.now(),
        'aDouble': 0.0,
      });
    }
    RealmObject.set(this, 'id', id);
    RealmObject.set(this, 'aBool', aBool);
    RealmObject.set(this, 'another', another);
    RealmObject.set(this, 'data', data);
    RealmObject.set(this, 'tidspunkt', timestamp);
    RealmObject.set(this, 'aDouble', aDouble);
    RealmObject.set(this, 'foo', foo);
    RealmObject.set(this, 'anOptionalString', anOptionalString);
    RealmObject.set(this, 'objectId', objectId);
    RealmObject.set<RealmList<int>>(this, 'list', RealmList<int>(list));
  }

  Bar._();

  @override
  String get id => RealmObject.get<String>(this, 'id') as String;
  @override
  set id(String value) => RealmObject.set(this, 'id', value);

  @override
  bool get aBool => RealmObject.get<bool>(this, 'aBool') as bool;
  @override
  set aBool(bool value) => RealmObject.set(this, 'aBool', value);

  @override
  bool get another => RealmObject.get<bool>(this, 'another') as bool;
  @override
  set another(bool value) => RealmObject.set(this, 'another', value);

  @override
  Uint8List get data => RealmObject.get<Uint8List>(this, 'data') as Uint8List;
  @override
  set data(Uint8List value) => RealmObject.set(this, 'data', value);

  @override
  DateTime get timestamp =>
      RealmObject.get<DateTime>(this, 'tidspunkt') as DateTime;
  @override
  set timestamp(DateTime value) => RealmObject.set(this, 'tidspunkt', value);

  @override
  double get aDouble => RealmObject.get<double>(this, 'aDouble') as double;
  @override
  set aDouble(double value) => RealmObject.set(this, 'aDouble', value);

  @override
  Foo? get foo => RealmObject.get<Foo>(this, 'foo') as Foo?;
  @override
  set foo(covariant Foo? value) => RealmObject.set(this, 'foo', value);

  @override
  RealmList<int> get list =>
      RealmObject.get<int>(this, 'list') as RealmList<int>;
  @override
  set list(covariant RealmList<int> value) => throw RealmUnsupportedSetError();

  @override
  String? get anOptionalString =>
      RealmObject.get<String>(this, 'anOptionalString') as String?;
  @override
  set anOptionalString(String? value) =>
      RealmObject.set(this, 'anOptionalString', value);

  @override
  ObjectId get objectId =>
      RealmObject.get<ObjectId>(this, 'objectId') as ObjectId;
  @override
  set objectId(ObjectId value) => RealmObject.set(this, 'objectId', value);

  @override
  Stream<RealmObjectChanges<Bar>> get changes =>
      RealmObject.getChanges<Bar>(this);

  @override
  Bar freeze() => RealmObject.freezeObject<Bar>(this);

  static SchemaObject get schema => _schema ??= _initSchema();
  static SchemaObject? _schema;
  static SchemaObject _initSchema() {
    RealmObject.registerFactory(Bar._);
    return const SchemaObject(Bar, 'Bar', [
      SchemaProperty('id', RealmPropertyType.string, primaryKey: true),
      SchemaProperty('aBool', RealmPropertyType.bool),
      SchemaProperty('another', RealmPropertyType.bool),
      SchemaProperty('data', RealmPropertyType.binary),
      SchemaProperty('tidspunkt', RealmPropertyType.timestamp,
          mapTo: 'tidspunkt'),
      SchemaProperty('aDouble', RealmPropertyType.double),
      SchemaProperty('foo', RealmPropertyType.object,
          optional: true, linkTarget: 'MyFoo'),
      SchemaProperty('list', RealmPropertyType.int,
          collectionType: RealmCollectionType.list),
      SchemaProperty('anOptionalString', RealmPropertyType.string,
          optional: true),
      SchemaProperty('objectId', RealmPropertyType.objectid),
    ]);
  }
}

class PrimitiveTypes extends _PrimitiveTypes with RealmEntity, RealmObject {
  PrimitiveTypes(
    String stringProp,
    bool boolProp,
    DateTime dateProp,
    double doubleProp,
    ObjectId objectIdProp,
  ) {
    RealmObject.set(this, 'stringProp', stringProp);
    RealmObject.set(this, 'boolProp', boolProp);
    RealmObject.set(this, 'dateProp', dateProp);
    RealmObject.set(this, 'doubleProp', doubleProp);
    RealmObject.set(this, 'objectIdProp', objectIdProp);
  }

  PrimitiveTypes._();

  @override
  String get stringProp =>
      RealmObject.get<String>(this, 'stringProp') as String;
  @override
  set stringProp(String value) => RealmObject.set(this, 'stringProp', value);

  @override
  bool get boolProp => RealmObject.get<bool>(this, 'boolProp') as bool;
  @override
  set boolProp(bool value) => RealmObject.set(this, 'boolProp', value);

  @override
  DateTime get dateProp =>
      RealmObject.get<DateTime>(this, 'dateProp') as DateTime;
  @override
  set dateProp(DateTime value) => RealmObject.set(this, 'dateProp', value);

  @override
  double get doubleProp =>
      RealmObject.get<double>(this, 'doubleProp') as double;
  @override
  set doubleProp(double value) => RealmObject.set(this, 'doubleProp', value);

  @override
  ObjectId get objectIdProp =>
      RealmObject.get<ObjectId>(this, 'objectIdProp') as ObjectId;
  @override
  set objectIdProp(ObjectId value) =>
      RealmObject.set(this, 'objectIdProp', value);

  @override
  Stream<RealmObjectChanges<PrimitiveTypes>> get changes =>
      RealmObject.getChanges<PrimitiveTypes>(this);

  @override
  PrimitiveTypes freeze() => RealmObject.freezeObject<PrimitiveTypes>(this);

  static SchemaObject get schema => _schema ??= _initSchema();
  static SchemaObject? _schema;
  static SchemaObject _initSchema() {
    RealmObject.registerFactory(PrimitiveTypes._);
    return const SchemaObject(PrimitiveTypes, 'PrimitiveTypes', [
      SchemaProperty('stringProp', RealmPropertyType.string),
      SchemaProperty('boolProp', RealmPropertyType.bool),
      SchemaProperty('dateProp', RealmPropertyType.timestamp),
      SchemaProperty('doubleProp', RealmPropertyType.double),
      SchemaProperty('objectIdProp', RealmPropertyType.objectid),
    ]);
  }
}